apps/src/gamelab/GameLab.js

```diff
@@ -0,0 +1,574 @@
+  if (!this.studioApp_) {
+    throw new Error("GameLab requires a StudioApp");
+  }
+
+  this.skin = config.skin;
+  this.level = config.level;
+
+  config.dropletConfig = dropletConfig;
+
+  var iconPath = 'media/turtle/' + (config.isLegacyShare && config.hideSource ? 'icons_white.png' : 'icons.png');
+
+  var showFinishButton = !this.level.isProjectLevel;
+  var finishButtonFirstLine = _.isEmpty(this.level.softButtons);
+  var firstControlsRow = require('./controls.html.ejs')({
+    assetUrl: this.studioApp_.assetUrl,
+    finishButton: finishButtonFirstLine && showFinishButton
+  });
+  var extraControlRows = require('./extraControlRows.html.ejs')({
+    assetUrl: this.studioApp_.assetUrl,
+    finishButton: !finishButtonFirstLine && showFinishButton
+  });
+
+  config.html = page({
+    assetUrl: this.studioApp_.assetUrl,
+    data: {
+      visualization: '',
+      localeDirection: this.studioApp_.localeDirection(),
+      controls: firstControlsRow,
+      extraControlRows: extraControlRows,
+      blockUsed : undefined,
+      idealBlockNumber : undefined,
+      editCode: this.level.editCode,
+      blockCounterClass : 'block-counter-default',
+      readonlyWorkspace: config.readonlyWorkspace
+    }
+  });
+
+  config.loadAudio = _.bind(this.loadAudio_, this);
+  config.afterInject = _.bind(this.afterInject_, this, config);
+
+  this.studioApp_.init(config);
+};
+
+GameLab.prototype.loadAudio_ = function () {
+  this.studioApp_.loadAudio(this.skin.winSound, 'win');
+  this.studioApp_.loadAudio(this.skin.startSound, 'start');
+  this.studioApp_.loadAudio(this.skin.failureSound, 'failure');
+};
+
+/**
+ * Code called after the blockly div + blockly core is injected into the document
+ */
+GameLab.prototype.afterInject_ = function (config) {
+
+  if (this.studioApp_.isUsingBlockly()) {
+    // Add to reserved word list: API, local variables in execution evironment
+    // (execute) and the infinite loop detection function.
+    Blockly.JavaScript.addReservedWords('GameLab,code');
+  }
+
+  // TODO (cpirich): remove?
+
+  // Adjust visualizationColumn width.
+  var visualizationColumn = document.getElementById('visualizationColumn');
+  visualizationColumn.style.width = '400px';
+};
+
+
+/**
+ * Reset GameLab to its initial state.
+ * @param {boolean} ignore Required by the API but ignored by this
+ *     implementation.
+ */
+GameLab.prototype.reset = function (ignore) {
+
+  this.eventHandlers = [];
+  window.clearInterval(this.tickIntervalId);
+  this.tickIntervalId = 0;
+  this.tickCount = 0;
+
+  // Discard the interpreter.
+  this.JSInterpreter = null;
+  this.executionError = null;
+};
+
+/**
+ * Click the run button.  Start the program.
+ */
+GameLab.prototype.runButtonClick = function () {
+  this.studioApp_.toggleRunReset('reset');
+  // document.getElementById('spinner').style.visibility = 'visible';
+  if (this.studioApp_.isUsingBlockly()) {
+    Blockly.mainBlockSpace.traceOn(true);
+  }
+  this.studioApp_.attempts++;
```

Comment:
Not relevant to this PR, but this increment feels like something that shouldn't be rewritten for every app type.